sierra2mlir/src/libfuncs/mod.rs

```diff
@@ -460,4 +463,69 @@ impl<'ctx> Compiler<'ctx> {
             arg,
         );
     }
+
+    pub fn create_libfunc_upcast(
+        &self,
+        func_decl: &LibfuncDeclaration,
+        parent_block: BlockRef<'ctx>,
+        storage: &mut Storage<'ctx>,
+    ) -> Result<()> {
+        let id = Self::normalize_func_name(func_decl.id.debug_name.as_deref().unwrap()).to_string();
+
+        let src_sierra_type = storage
+            .types
+            .get(&match &func_decl.long_id.generic_args[0] {
+                GenericArg::Type(x) => x.id.to_string(),
+                _ => todo!("invalid generic kind"),
+            })
+            .expect("type to exist");
+        let dst_sierra_type = storage
+            .types
+            .get(&match &func_decl.long_id.generic_args[1] {
+                GenericArg::Type(x) => x.id.to_string(),
+                _ => todo!("invalid generic kind"),
+            })
+            .expect("type to exist");
+
+        let src_type = src_sierra_type.get_type();
+        let dst_type = dst_sierra_type.get_type();
+
+        match src_type
+            .get_width()
+            .unwrap()
+            .cmp(&dst_type.get_width().unwrap())
+        {
+            Ordering::Less => {
+                let region = Region::new();
+                let block = Block::new(&[(src_type, Location::unknown(&self.context))]);
+
+                let op_ref = self.op_zext(&block, block.argument(0)?.into(), dst_type);
+
+                self.op_return(&block, &[op_ref.result(0)?.into()]);
+                region.append_block(block);
+
+                let func = self.op_func(
+                    &id,
+                    &format!("({src_type}) -> {dst_type}"),
```

Comment:
just to be consistent, its probably better to use `create_fn_signature` here

Reply:
Oh I just "copy-adjust-pasted" from a different libfunc. I'll change both.